### ISSUE TITLE
Use Singularity?

### ISSUE THREAD

**Issue by photocyte** [transXpress/transXpress-nextflow#20]
Singularity is a virtualization layer, like Docker, which is friendly with HPC systems.

At least on my current cluster the core libraries (glib, kernel, etc.), are a bit aged, and some software (signalp5.0b) which does not specify dependencies is not working properly out of the box.  I haven't tried deeploc, but I imagine it might be similar.

An alternative solution would be to build Singularity containers which can run particular pieces of software - then the software is isolated from these core dependencies, except necessitating that Singularity must be installed.  This node shows an example of building a Singularity container within a nextflow process:
https://github.com/transXpress/transXpress-nextflow/blob/dc0bf9eeef306cf22704494a55160f574d2ff37d/post_assembly_pipelines/assess-viral-diversity.nf#L414 

One issue is that the SignalP5.0b license is a bit restrictive, unsure if it can be distributed in a prebuilt Singularity container.  An interesting angle is that Singularity containers can now be encrypted https://insidehpc.com/2019/09/singularity-3-4-0-enables-build-and-run-encrypted-containers/ , which might allow for them to be distributed.

Can also run the whole pipeline in Singularity (using nextflow run <your script> -with-singularity [singularity image file]).  Notably, singularity-hub.org allows for easy building of Singularity images.  E.g. the "Singularity" build definition file in the current repository: https://github.com/transXpress/transXpress-nextflow/blob/dc0bf9eeef306cf22704494a55160f574d2ff37d/Singularity#L1, currently contains Conda + all the conda installs, is currently linked to singularity-hub, and a singularity image was successfully built by singularity-hub (can pull with `singularity pull shub://photocyte/transXpress-nextflow`)

So in short, one can trade the Conda dependency, for a Singularity dependency, and use that to manage the dependencies.  Singularity however requires root access for installation, but ultimately should be more flexible and be better at software isolation than Conda.

**Comment by tomas-pluskal** [comment]
I think both signalp and deeploc have strict licenses and need to be downloaded and installed separately. That makes the container approach less attractive. (TODO for my own lab: develop a better and free alternative to signalp/deeploc)
